### ISSUE TITLE
Git submodules fetch from public repo fails if SSH key is used for private repo

### ISSUE THREAD

**Issue by logicfox** [argoproj/argo#1498]
**Is this a BUG REPORT or FEATURE REQUEST?**:
BUG REPORT
**What happened**:
I have a private Bitbucket repo with SSH access that have a submodule from a public Github repo. Setting up a Git Artifact to use SSH Private Key to pull fails when it tries to pull the public submodule with the following error:

```
k -n argo logs argo-ci-2611739299 init
time="2019-07-25T14:44:37Z" level=info msg="Creating a docker executor"
time="2019-07-25T14:44:37Z" level=info msg="Executor (version: v2.3.0, build_date: 2019-05-20T22:10:54Z) initialized (pod: argo/argo-ci-2611739299) with template:\n{\"name\":\"checkout\",\"inputs\":{\"artifacts\":[{\"name\":\"source\",\"path\":\"/src\",\"git\":{\"repo\":\"git@bitbucket.org:private-repo/example-cicd.git\",\"revision\":\"8980926c17a243dde19fcf99db5ecfb6622057ad\",\"sshPrivateKeySecret\":{\"name\":\"bitbucketssh\",\"key\":\"ssh-ci\"}}}]},\"outputs\":{\"artifacts\":[{\"name\":\"source\",\"path\":\"/src\"}]},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"private.azurecr.io/alpine-swiss-army-knife:latest\",\"command\":[\"/bin/sh\",\"-c\"],\"args\":[\"cat /var/run/secrets/gpg/jenkins.pgp | gpg --import \\u0026\\u0026 git-crypt unlock \\u0026\\u0026 git status \\u0026\\u0026 ls -l\"],\"workingDir\":\"/src\",\"resources\":{},\"volumeMounts\":[{\"name\":\"gpgkey\",\"mountPath\":\"/var/run/secrets/gpg\"}]},\"archiveLocation\":{\"archiveLogs\":true,\"s3\":{\"endpoint\":\"minio.argo.svc.cluster.local:9000\",\"bucket\":\"argo-artifacts\",\"insecure\":true,\"accessKeySecret\":{\"name\":\"minio-azure-blob-secrets\",\"key\":\"accesskey\"},\"secretKeySecret\":{\"name\":\"minio-azure-blob-secrets\",\"key\":\"secretkey\"},\"key\":\"argo-ci/argo-ci-2611739299\"}}}"
time="2019-07-25T14:44:37Z" level=info msg="Start loading input artifacts..."
time="2019-07-25T14:44:37Z" level=info msg="Downloading artifact: source"
time="2019-07-25T14:44:39Z" level=error msg="executor error: invalid auth method\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapError\n\t/go/src/github.com/argoproj/argo/errors/errors.go:71\ngithub.com/argoproj/argo/workflow/artifacts/git.gitClone\n\t/go/src/github.com/argoproj/argo/workflow/artifacts/git/git.go:91\ngithub.com/argoproj/argo/workflow/artifacts/git.(*GitArtifactDriver).Load\n\t/go/src/github.com/argoproj/argo/workflow/artifacts/git/git.go:41\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).LoadArtifacts\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:165\ngithub.com/argoproj/argo/cmd/argoexec/commands.loadArtifacts\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/init.go:34\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewInitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/init.go:14\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
time="2019-07-25T14:44:39Z" level=info msg="Alloc=5542 TotalAlloc=13988 Sys=70590 NumGC=5 Goroutines=3"
time="2019-07-25T14:44:39Z" level=fatal msg="invalid auth method\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapError\n\t/go/src/github.com/argoproj/argo/errors/errors.go:71\ngithub.com/argoproj/argo/workflow/artifacts/git.gitClone\n\t/go/src/github.com/argoproj/argo/workflow/artifacts/git/git.go:91\ngithub.com/argoproj/argo/workflow/artifacts/git.(*GitArtifactDriver).Load\n\t/go/src/github.com/argoproj/argo/workflow/artifacts/git/git.go:41\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).LoadArtifacts\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:165\ngithub.com/argoproj/argo/cmd/argoexec/commands.loadArtifacts\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/init.go:34\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewInitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/init.go:14\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
```

**What you expected to happen**:
Git Artifact should checkout the repo and its submodules irrespective of where they are hosted. Private repos should be authenticated against and public submodules should be authenticated only if the key for the repo is provided.

**How to reproduce it (as minimally and precisely as possible)**:
- Set up a private repo with Bitbucket/Gitlab/self-hosted with SSH access. 
- Add a public Github repo as submodule
- Set up a workflow to checkout the git artifact with the SSH private key
- Check logs of the init container of the checkout step pod 

**Anything else we need to know?**:
The issue can be mitigated by providing a way to specify a host-sshkey map in the Git artifact step and use that to generate `~/.ssh/config` like this SO example: https://superuser.com/a/272613

**Environment**:
- Argo version: 2.3.0
- Kubernetes version : AKS 
```
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.5", GitCommit:"2166946f41b36dea2c4626f90a77706f426cdea2", GitTreeState:"clean", BuildDate:"2019-03-25T15:19:22Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Other debugging information (if applicable)**:
- workflow result:
```
argo -n argo get argo-ci
Name:                argo-ci
Namespace:           argo
ServiceAccount:      default
Status:              Failed
Message:             child 'argo-ci-2611739299' failed
Created:             Thu Jul 25 16:44:34 +0200 (9 minutes ago)
Started:             Thu Jul 25 16:44:34 +0200 (9 minutes ago)
Finished:            Thu Jul 25 16:44:40 +0200 (9 minutes ago)
Duration:            6 seconds
Parameters:
  repo:              git@bitbucket.org:private-repo/example-cicd.git
  branch:            master
  commit-id:         8980926c17a243dde19fcf99db5ecfb6622057ad

STEP             PODNAME             DURATION  MESSAGE
 ✖ argo-ci                                     child 'argo-ci-2611739299' failed
 └---⚠ checkout  argo-ci-2611739299  5s        failed to load artifacts
```
- executor logs:
```
$ k -n argo logs argo-ci-2611739299 init
time="2019-07-25T14:44:37Z" level=info msg="Creating a docker executor"
time="2019-07-25T14:44:37Z" level=info msg="Executor (version: v2.3.0, build_date: 2019-05-20T22:10:54Z) initialized (pod: argo/argo-ci-2611739299) with template:\n{\"name\":\"checkout\",\"inputs\":{\"artifacts\":[{\"name\":\"source\",\"path\":\"/src\",\"git\":{\"repo\":\"git@bitbucket.org:private-repo/example-cicd.git\",\"revision\":\"8980926c17a243dde19fcf99db5ecfb6622057ad\",\"sshPrivateKeySecret\":{\"name\":\"bitbucketssh\",\"key\":\"ssh-ci\"}}}]},\"outputs\":{\"artifacts\":[{\"name\":\"source\",\"path\":\"/src\"}]},\"metadata\":{},\"container\":{\"name\":\"\",\"image\":\"private.azurecr.io/alpine-swiss-army-knife:latest\",\"command\":[\"/bin/sh\",\"-c\"],\"args\":[\"cat /var/run/secrets/gpg/jenkins.pgp | gpg --import \\u0026\\u0026 git-crypt unlock \\u0026\\u0026 git status \\u0026\\u0026 ls -l\"],\"workingDir\":\"/src\",\"resources\":{},\"volumeMounts\":[{\"name\":\"gpgkey\",\"mountPath\":\"/var/run/secrets/gpg\"}]},\"archiveLocation\":{\"archiveLogs\":true,\"s3\":{\"endpoint\":\"minio.argo.svc.cluster.local:9000\",\"bucket\":\"argo-artifacts\",\"insecure\":true,\"accessKeySecret\":{\"name\":\"minio-azure-blob-secrets\",\"key\":\"accesskey\"},\"secretKeySecret\":{\"name\":\"minio-azure-blob-secrets\",\"key\":\"secretkey\"},\"key\":\"argo-ci/argo-ci-2611739299\"}}}"
time="2019-07-25T14:44:37Z" level=info msg="Start loading input artifacts..."
time="2019-07-25T14:44:37Z" level=info msg="Downloading artifact: source"
time="2019-07-25T14:44:39Z" level=error msg="executor error: invalid auth method\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapError\n\t/go/src/github.com/argoproj/argo/errors/errors.go:71\ngithub.com/argoproj/argo/workflow/artifacts/git.gitClone\n\t/go/src/github.com/argoproj/argo/workflow/artifacts/git/git.go:91\ngithub.com/argoproj/argo/workflow/artifacts/git.(*GitArtifactDriver).Load\n\t/go/src/github.com/argoproj/argo/workflow/artifacts/git/git.go:41\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).LoadArtifacts\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:165\ngithub.com/argoproj/argo/cmd/argoexec/commands.loadArtifacts\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/init.go:34\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewInitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/init.go:14\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
time="2019-07-25T14:44:39Z" level=info msg="Alloc=5542 TotalAlloc=13988 Sys=70590 NumGC=5 Goroutines=3"
time="2019-07-25T14:44:39Z" level=fatal msg="invalid auth method\ngithub.com/argoproj/argo/errors.Wrap\n\t/go/src/github.com/argoproj/argo/errors/errors.go:88\ngithub.com/argoproj/argo/errors.InternalWrapError\n\t/go/src/github.com/argoproj/argo/errors/errors.go:71\ngithub.com/argoproj/argo/workflow/artifacts/git.gitClone\n\t/go/src/github.com/argoproj/argo/workflow/artifacts/git/git.go:91\ngithub.com/argoproj/argo/workflow/artifacts/git.(*GitArtifactDriver).Load\n\t/go/src/github.com/argoproj/argo/workflow/artifacts/git/git.go:41\ngithub.com/argoproj/argo/workflow/executor.(*WorkflowExecutor).LoadArtifacts\n\t/go/src/github.com/argoproj/argo/workflow/executor/executor.go:165\ngithub.com/argoproj/argo/cmd/argoexec/commands.loadArtifacts\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/init.go:34\ngithub.com/argoproj/argo/cmd/argoexec/commands.NewInitCommand.func1\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/commands/init.go:14\ngithub.com/spf13/cobra.(*Command).execute\n\t/go/src/github.com/spf13/cobra/command.go:766\ngithub.com/spf13/cobra.(*Command).ExecuteC\n\t/go/src/github.com/spf13/cobra/command.go:852\ngithub.com/spf13/cobra.(*Command).Execute\n\t/go/src/github.com/spf13/cobra/command.go:800\nmain.main\n\t/go/src/github.com/argoproj/argo/cmd/argoexec/main.go:17\nruntime.main\n\t/usr/local/go/src/runtime/proc.go:201\nruntime.goexit\n\t/usr/local/go/src/runtime/asm_amd64.s:1333"
$ kubectl logs <failedpodname> -c wait
```
- workflow-controller logs:
```
time="2019-07-25T14:44:34Z" level=info msg="Processing workflow" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:34Z" level=info msg="Updated phase  -> Running" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:34Z" level=info msg="Steps node argo-ci (argo-ci) initialized Running" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:34Z" level=info msg="StepGroup node argo-ci[0] (argo-ci-1654851219) initialized Running" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:34Z" level=info msg="Created pod: argo-ci[0].checkout (argo-ci-2611739299)" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:34Z" level=info msg="Pod node argo-ci[0].checkout (argo-ci-2611739299) initialized Pending" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:34Z" level=info msg="Workflow step group node argo-ci[0] (argo-ci-1654851219) not yet completed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:34Z" level=info msg="Workflow update successful" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:35Z" level=info msg="Processing workflow" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:35Z" level=info msg="Updating node argo-ci[0].checkout (argo-ci-2611739299) message: PodInitializing"
time="2019-07-25T14:44:35Z" level=info msg="Workflow step group node argo-ci[0] (argo-ci-1654851219) not yet completed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:35Z" level=info msg="Workflow update successful" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:36Z" level=info msg="Processing workflow" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:36Z" level=info msg="Workflow step group node argo-ci[0] (argo-ci-1654851219) not yet completed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:37Z" level=info msg="Processing workflow" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:37Z" level=info msg="Workflow step group node argo-ci[0] (argo-ci-1654851219) not yet completed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Processing workflow" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Updating node argo-ci[0].checkout (argo-ci-2611739299) status Pending -> Error"
time="2019-07-25T14:44:40Z" level=info msg="Updating node argo-ci[0].checkout (argo-ci-2611739299) message: failed to load artifacts"
time="2019-07-25T14:44:40Z" level=info msg="Step group node argo-ci[0] (argo-ci-1654851219) deemed failed: child 'argo-ci-2611739299' failed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="node argo-ci[0] (argo-ci-1654851219) phase Running -> Failed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="node argo-ci[0] (argo-ci-1654851219) message: child 'argo-ci-2611739299' failed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="node argo-ci[0] (argo-ci-1654851219) finished: 2019-07-25 14:44:40.516397095 +0000 UTC" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="step group argo-ci-1654851219 was unsuccessful: child 'argo-ci-2611739299' failed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Outbound nodes of argo-ci-2611739299 is [argo-ci-2611739299]" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Outbound nodes of argo-ci is [argo-ci-2611739299]" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="node argo-ci (argo-ci) phase Running -> Failed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="node argo-ci (argo-ci) message: child 'argo-ci-2611739299' failed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="node argo-ci (argo-ci) finished: 2019-07-25 14:44:40.516445594 +0000 UTC" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Checking daemoned children of argo-ci" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Updated phase Running -> Failed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Updated message  -> child 'argo-ci-2611739299' failed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Marking workflow completed" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Checking daemoned children of " namespace=argo workflow=argo-ci
time="2019-07-25T14:44:40Z" level=info msg="Workflow update successful" namespace=argo workflow=argo-ci
time="2019-07-25T14:44:41Z" level=info msg="Labeled pod argo/argo-ci-2611739299 completed"
time="2019-07-25T14:46:20Z" level=info msg="Alloc=24537 TotalAlloc=39556545 Sys=73022 NumGC=2671 Goroutines=68"
```

**Comment by logicfox** [comment]
This issue is still relevant

**Comment by alexec** [comment]
This may have been fixed in v2.10. Can you please try with `:latest`?
